File: python/update_birthdate.py
import argparse
import os
import random
import sys
import time
import uuid
from datetime import datetime
from pathlib import Path

import requests
from dotenv import load_dotenv

load_dotenv(dotenv_path=Path(__file__).parent.parent / '.env')

KEY = os.environ.get('SCHOOLOGY_KEY')
SECRET = os.environ.get('SCHOOLOGY_SECRET')
STUDENT_ROLE_ID = os.environ.get('STUDENT_ROLE_ID')

if not all([KEY, SECRET, STUDENT_ROLE_ID]):
    print('Missing required environment variables: SCHOOLOGY_KEY, SCHOOLOGY_SECRET, STUDENT_ROLE_ID')
    sys.exit(1)

CURRENT_YEAR = datetime.now().year
TARGET_DATE = f'{CURRENT_YEAR}-01-01'
BASE_URL = 'https://api.schoology.com'


def oauth_header():
    nonce = uuid.uuid4().hex
    timestamp = int(time.time())
    return (
        f'OAuth realm="Schoology API",'
        f'oauth_consumer_key="{KEY}",'
        f'oauth_token="",'
        f'oauth_nonce="{nonce}",'
        f'oauth_timestamp="{timestamp}",'
        f'oauth_signature_method="PLAINTEXT",'
        f'oauth_signature="{SECRET}%26",'
        f'oauth_version="1.0"'
    )


def api_request(method, path, json=None):
    url = f'{BASE_URL}{path}'
    headers = {
        'Authorization': oauth_header(),
        'Content-Type': 'application/json',
        'Accept': 'application/json',
    }
    response = requests.request(method, url, headers=headers, json=json)
    response.raise_for_status()
    return response.json() if response.content else {}


def fetch_all_students():
    students = []
    start = 0
    limit = 200

    while True:
        path = f'/v1/users?role_ids={STUDENT_ROLE_ID}&limit={limit}&start={start}'
        data = api_request('GET', path)
        page = data.get('user', [])
        students.extend(page)
        total = int(data.get('total', 0))
        start += len(page)
        if not page or start >= total:
            break

    return students


def fetch_student_by_uid(school_uid):
    from urllib.parse import quote
    data = api_request('GET', f'/v1/users?school_uid={quote(str(school_uid))}')
    users = data.get('user', [])
    if not users:
        print(f'No user found with school_uid: {school_uid}')
        sys.exit(1)
    return users[0]


def update_students(students):
    to_update = [s for s in students if s.get('birthday_date') != TARGET_DATE]
    skipped = len(students) - len(to_update)

    print(f'Checked:  {len(students)}')
    print(f'Skipped:  {skipped} (already {TARGET_DATE})')
    print(f'Updating: {len(to_update)}')

    if not to_update:
        print('Nothing to do.')
        return

    updated = 0
    for i in range(0, len(to_update), 50):
        batch = to_update[i:i + 50]
        payload = {
            'users': {
                'user': [{'id': s['id'], 'birthday_date': TARGET_DATE} for s in batch]
            }
        }
        api_request('PUT', '/v1/users', json=payload)
        updated += len(batch)
        print(f'Updated {updated}/{len(to_update)}')
        if updated < len(to_update):
            time.sleep(0.5)

    print('Done.')


def main():
    parser = argparse.ArgumentParser(
        prog='python update_birthdate.py',
        description='Anonymize student birthday_date fields in Schoology.',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog=(
            'examples:\n'
            '  python update_birthdate.py --uid 100234\n'
            '  python update_birthdate.py --random\n'
            '  python update_birthdate.py --all'
        ),
    )
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument('--all', action='store_true', help='Update all student accounts')
    group.add_argument('--uid', metavar='school_uid', help='Update one specific student by school_uid')
    group.add_argument('--random', action='store_true', help='Update one randomly selected student (smoke test)')

    args = parser.parse_args()

    if args.uid:
        print(f'Fetching student with school_uid: {args.uid}...')
        student = fetch_student_by_uid(args.uid)
        print(f'Found: id {student["id"]}')
        update_students([student])

    elif args.random:
        print('Fetching one page of students to pick from...')
        data = api_request('GET', f'/v1/users?role_ids={STUDENT_ROLE_ID}&limit=200&start=0')
        students = data.get('user', [])
        if not students:
            print('No students found.')
            sys.exit(1)
        student = random.choice(students)
        print(f'Randomly selected: id {student["id"]}')
        update_students([student])

    elif args.all:
        print('Fetching all students...')
        students = fetch_all_students()
        print(f'Total fetched: {len(students)}')
        update_students(students)


if __name__ == '__main__':
    main()

File: php/update_birthdate.php
<?php

$usage = <<<HELP
Usage: php update_birthdate.php <mode>

Modes:
  --all              Update all student accounts
  --uid <school_uid> Update one specific student by school_uid
  --random           Update one randomly selected student (smoke test)

Examples:
  php update_birthdate.php --uid 100234
  php update_birthdate.php --random
  php update_birthdate.php --all
HELP;

// Parse arguments
$mode    = $argv[1] ?? null;
$modeArg = $argv[2] ?? null;

if (!$mode || in_array($mode, ['--help', '-h'])) {
    echo $usage . "\n";
    exit(0);
}

if (!in_array($mode, ['--all', '--uid', '--random'])) {
    fwrite(STDERR, "Unknown option: {$mode}\n\n");
    echo $usage . "\n";
    exit(1);
}

if ($mode === '--uid' && !$modeArg) {
    fwrite(STDERR, "--uid requires a school_uid argument\n\n");
    echo $usage . "\n";
    exit(1);
}

// Load .env
$envFile = __DIR__ . '/../.env';
if (file_exists($envFile)) {
    foreach (file($envFile, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES) as $line) {
        if (strpos(trim($line), '#') === 0) continue;
        [$name, $value] = array_map('trim', explode('=', $line, 2));
        putenv("$name=$value");
        $_ENV[$name] = $value;
    }
}

$key           = getenv('SCHOOLOGY_KEY');
$secret        = getenv('SCHOOLOGY_SECRET');
$studentRoleId = getenv('STUDENT_ROLE_ID');

if (!$key || !$secret || !$studentRoleId) {
    fwrite(STDERR, "Missing required environment variables: SCHOOLOGY_KEY, SCHOOLOGY_SECRET, STUDENT_ROLE_ID\n");
    exit(1);
}

$currentYear = (int) date('Y');
$targetDate  = "{$currentYear}-01-01";
$baseUrl     = 'https://api.schoology.com';

function oauthHeader(string $key, string $secret): string
{
    $nonce     = bin2hex(random_bytes(16));
    $timestamp = time();
    return implode(',', [
        'OAuth realm="Schoology API"',
        "oauth_consumer_key=\"{$key}\"",
        'oauth_token=""',
        "oauth_nonce=\"{$nonce}\"",
        "oauth_timestamp=\"{$timestamp}\"",
        'oauth_signature_method="PLAINTEXT"',
        "oauth_signature=\"{$secret}%26\"",
        'oauth_version="1.0"',
    ]);
}

function apiRequest(string $method, string $path, ?array $body, string $key, string $secret): array
{
    global $baseUrl;
    $url = $baseUrl . $path;
    $ch  = curl_init($url);
    $headers = [
        'Authorization: ' . oauthHeader($key, $secret),
        'Content-Type: application/json',
        'Accept: application/json',
    ];
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_HTTPHEADER, $headers);
    curl_setopt($ch, CURLOPT_CUSTOMREQUEST, $method);
    if ($body !== null) {
        curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($body));
    }
    $response = curl_exec($ch);
    $status   = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    curl_close($ch);
    if ($status < 200 || $status >= 300) {
        throw new RuntimeException("HTTP {$status}: {$response}");
    }
    return $response ? json_decode($response, true) : [];
}

function fetchAllStudents(string $roleId, string $key, string $secret): array
{
    $students = [];
    $start    = 0;
    $limit    = 200;
    while (true) {
        $path = "/v1/users?role_ids={$roleId}&limit={$limit}&start={$start}";
        $data = apiRequest('GET', $path, null, $key, $secret);
        $page = $data['user'] ?? [];
        foreach ($page as $user) { $students[] = $user; }
        $total  = (int) ($data['total'] ?? 0);
        $start += count($page);
        if (empty($page) || $start >= $total) break;
    }
    return $students;
}

function fetchStudentByUid(string $uid, string $key, string $secret): array
{
    $data  = apiRequest('GET', '/v1/users?school_uid=' . urlencode($uid), null, $key, $secret);
    $users = $data['user'] ?? [];
    if (empty($users)) {
        fwrite(STDERR, "No user found with school_uid: {$uid}\n");
        exit(1);
    }
    return $users[0];
}

function updateStudents(array $students, string $targetDate, string $key, string $secret): void
{
    $toUpdate = array_values(array_filter($students, fn($s) => ($s['birthday_date'] ?? '') !== $targetDate));
    $skipped  = count($students) - count($toUpdate);

    echo "Checked:  " . count($students) . "\n";
    echo "Skipped:  {$skipped} (already {$targetDate})\n";
    echo "Updating: " . count($toUpdate) . "\n";

    if (empty($toUpdate)) {
        echo "Nothing to do.\n";
        return;
    }

    $updated = 0;
    foreach (array_chunk($toUpdate, 50) as $batch) {
        $payload = [
            'users' => [
                'user' => array_map(fn($s) => ['id' => $s['id'], 'birthday_date' => $targetDate], $batch),
            ],
        ];
        apiRequest('PUT', '/v1/users', $payload, $key, $secret);
        $updated += count($batch);
        echo "Updated {$updated}/" . count($toUpdate) . "\n";
        if ($updated < count($toUpdate)) {
            usleep(500000);
        }
    }

    echo "Done.\n";
}

// Main
if ($mode === '--uid') {
    echo "Fetching student with school_uid: {$modeArg}...\n";
    $student = fetchStudentByUid($modeArg, $key, $secret);
    echo "Found: id {$student['id']}\n";
    updateStudents([$student], $targetDate, $key, $secret);

} elseif ($mode === '--random') {
    echo "Fetching one page of students to pick from...\n";
    $data     = apiRequest('GET', "/v1/users?role_ids={$studentRoleId}&limit=200&start=0", null, $key, $secret);
    $students = $data['user'] ?? [];
    if (empty($students)) {
        fwrite(STDERR, "No students found.\n");
        exit(1);
    }
    $student = $students[array_rand($students)];
    echo "Randomly selected: id {$student['id']}\n";
    updateStudents([$student], $targetDate, $key, $secret);

} elseif ($mode === '--all') {
    echo "Fetching all students...\n";
    $students = fetchAllStudents($studentRoleId, $key, $secret);
    echo 'Total fetched: ' . count($students) . "\n";
    updateStudents($students, $targetDate, $key, $secret);
}
